Add logbook and device trigger platforms to Shelly

thecode opened the pull request:
Add `logbook` platform for describing “shelly.click” event
Add `device_trigger` platform for adding automation based on click events:

Example of logbook event:
``` Shelly 'single' click event for Test I3 channel 3 was fired. ```
(Test I3 is the name of the device)

Example of automation triggers:
```
First button triple clicked
First button long clicked and then single clicked
First button double clicked
First button long clicked
First button single clicked and then long clicked
First button single clicked
Second button triple clicked
..
Second button single clicked
```
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

thecode committed:
Add logbook and device trigger platforms to Shelly

Add `logbook` platform for describing “shelly.click” event
Add `device_trigger` platform for adding automation based on click events:

Example of logbook event:
Shelly 'single' click event for Test I3 channel 3 was fired.
(Test I3 is the name of the device)

Example of automation triggers:
First button triple clicked
First button long clicked and then single clicked
First button double clicked
First button long clicked
First button single clicked and then long clicked
First button single clicked
Second button triple clicked
..
Second button single clicked

thecode committed:
Fix codespell

thecode committed:
Remove pylint added for debug

thecode committed:
Add tests

balloob commented on homeassistant/components/shelly/const.py:
```diff
@@ -32,3 +32,38 @@
+    "double_click",
+    "triple_click",
+    "long_click",
+    "single_long_click",
+    "long_single_click",
```
what's the difference?

thecode replied:
The order of the button press:
Single (short) click and then long click
Long click and then single click

balloob commented on homeassistant/components/shelly/device_trigger.py:
```diff
@@ -0,0 +1,110 @@
+TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_TYPE): vol.In(SUPPORTED_TRIGGER_TYPES),
+        vol.Required(CONF_SUBTYPE): vol.In(TRIGGER_SUBTYPES.keys()),
```
Keys is default iterator.

```suggestion
        vol.Required(CONF_SUBTYPE): vol.In(TRIGGER_SUBTYPES),
```

thecode replied:
Fixed

balloob commented on homeassistant/components/shelly/const.py:
```diff
@@ -32,3 +32,38 @@
+    "long_click",
+}
+
+SUPPORTED_TRIGGER_TYPES = SHIX3_1_TRIGGER_TYPES = {
```
Why are we not using same values as `INPUTS_EVENTS_DICT` ?

thecode replied:
I will use the same values (reached a conclusion that this would be the most clean solution and not adding another event type)

balloob commented on homeassistant/components/shelly/device_trigger.py:
```diff
@@ -0,0 +1,110 @@
+            event_trigger.CONF_EVENT_DATA: {
+                ATTR_DEVICE_ID: config[CONF_DEVICE_ID],
+                ATTR_CHANNEL: TRIGGER_SUBTYPES[config[CONF_SUBTYPE]],
+                ATTR_CLICK_TYPE: config[CONF_TYPE].split("_")[0],
```
This is weird and seems dangerous. Is this because we're using different values than `INPUTS_EVENTS_DICT` vs `CONF_TYPE` values ?

thecode replied:
I think about it again and will use the same `INPUTS_EVENTS_DICT`

balloob commented on homeassistant/components/shelly/utils.py:
```diff
@@ -35,54 +43,76 @@ def get_device_name(device: aioshelly.Device) -> str:
+    if block.type == "input":
+        # Shelly Dimmer/1L has two input channels and missing "num_inputs"
+        if device.settings["device"]["type"] in ["SHDM-1", "SHDM-2", "SHSW-L"]:
+            channels = 2
+        else:
+            channels = device.shelly.get("num_inputs")
+    elif block.type == "emeter":
+        channels = device.shelly.get("num_emeters")
+    elif block.type in ["relay", "light"]:
+        channels = device.shelly.get("num_outputs")
+    elif block.type in ["roller", "device"]:
+        channels = 1
+
+    return channels or 1
```
Just return as soon as you're done:

```suggestion
            return 2

        return device.shelly.get("num_inputs")
    if block.type == "emeter":
        return device.shelly.get("num_emeters")
    if block.type in ["relay", "light"]:
        return device.shelly.get("num_outputs")
    if block.type in ["roller", "device"]:
        return 1

    return 1
```

thecode replied:
Offline discussion,  since some devices does not return a value we use 1 instead so we leave the code as before the suggestion.

balloob commented on homeassistant/components/shelly/utils.py:
```diff
@@ -35,54 +43,76 @@ def get_device_name(device: aioshelly.Device) -> str:
 
+def get_number_of_channels(device: aioshelly.Device, block: aioshelly.Block) -> int:
+    """Get number of channels for block type."""
+    channels = None
```
```suggestion
```

balloob reviewed (commented):
Looks great. Almost there!

thecode committed:
Rebase

thecode committed:
Fix Rebase & Apply PR review suggestions

Fix tests after rebasing
Use `INPUTS_EVENTS_DICT` for input triggers
Apply PR suggestions
